File: src/api/taskcard.ts
import { invoke } from "@tauri-apps/api/core";
import type { Settings } from "./settings";

export interface TaskWebviewInterface {
  panel_name: string;
  interface_port: number;
  localhost_only: boolean;
}

export interface TaskVncInterface {
  panel_name: string;
}

export interface TaskCardConfig {
  id: string;
  name: string;
  env: Record<string, string>;
}

export interface TaskCardTask {
  uuid: string;
  uuid_conflict: boolean;
  id: string;
  prefix_path: string;
  name: string;
  description: string;
  workdir: string;
  command: string;
  env_count: number;
  configs: TaskCardConfig[];
  default_config?: string;
  running_config_id?: string;
  requires_sudo: boolean;
  webview_interface?: TaskWebviewInterface[];
  vnc_interface?: TaskVncInterface[];
  folder: string;
  status: "running" | "stopped";
  pid?: number;
  started_at_ms?: number;
  log_file?: string;
}

export interface TaskCardGroupTask {
  task: string;
  config?: string;
  wait_after_sec: number;
  env: Record<string, string>;
  prefix_path?: string;
}

export interface TaskCardGroup {
  version: string;
  uuid: string;
  id: string;
  prefix_path: string;
  name: string;
  description: string;
  tasks: TaskCardGroupTask[];
  folder: string;
}

export interface TaskCardSnapshot {
  root: string;
  default_route_ip: string;
  vnc_port: number;
  search_paths: string[];
  discovered_task_dirs: string[];
  discovered_group_dirs: string[];
  tasks: TaskCardTask[];
  groups: TaskCardGroup[];
  uuid_conflicts: UuidConflict[];
  errors: string[];
}

export interface UuidDefinitionRef {
  kind: "task" | "group";
  id: string;
  prefix_path: string;
  path: string;
}

export interface UuidConflict {
  uuid: string;
  definitions: UuidDefinitionRef[];
}

export interface ResearchResult {
  search_paths: string[];
  discovered_task_dirs: string[];
  discovered_group_dirs: string[];
}

export interface TaskLogSummary {
  file: string;
  task_id: string;
  config_id?: string;
  started_at_ms: number;
  modified_at_ms: number;
  bytes: number;
  active: boolean;
}

export interface TaskLogContent {
  file: string;
  content: string;
  truncated: boolean;
}

export interface TaskLogChunk {
  file: string;
  content: string;
  next_offset: number;
  reset: boolean;
}

export interface YamlDocument {
  content: string;
  folder: string;
}

export function fetchTaskCard() {
  return invoke<TaskCardSnapshot>("taskcard_snapshot");
}

export function researchTaskCard() {
  return invoke<ResearchResult>("taskcard_research");
}

export function addSearchPath(path: string) {
  return invoke<Settings>("taskcard_add_search_path", { path });
}

export interface PathSuggestions {
  query: string;
  paths: string[];
}

export function listPathSuggestions(prefix: string) {
  return invoke<PathSuggestions>("list_path_suggestions_command", { prefix });
}

export function removeSearchPath(path: string) {
  return invoke<Settings>("taskcard_remove_search_path", { path });
}

export function startTask(prefixPath: string, id: string, configId?: string, sudoPassword?: string) {
  return invoke<void>("taskcard_start_task", {
    prefixPath,
    id,
    configId: configId || null,
    sudoPassword: sudoPassword ?? null,
  });
}

export function stopTask(prefixPath: string, id: string) {
  return invoke<void>("taskcard_stop_task", { prefixPath, id });
}

export function restartTask(prefixPath: string, id: string, configId?: string, sudoPassword?: string) {
  return invoke<void>("taskcard_restart_task", {
    prefixPath,
    id,
    configId: configId || null,
    sudoPassword: sudoPassword ?? null,
  });
}

export function stopAllTasks() {
  return invoke<string[]>("taskcard_stop_all");
}

export interface ManagedProcessInfo {
  pid: number;
  ppid: number;
  name: string;
  command: string;
  state: string;
}

export interface ManagedProcessGroup {
  uuid: string;
  prefix_path: string;
  task_id: string;
  leader_pid: number;
  pgid: number;
  started_at_ms: number;
  log_file: string;
  config_id?: string | null;
  task_running: boolean;
  orphaned: boolean;
  processes: ManagedProcessInfo[];
}

export function fetchManagedProcesses() {
  return invoke<ManagedProcessGroup[]>("managed_processes");
}

export function stopManagedProcesses(uuid: string) {
  return invoke<void>("stop_managed_processes", { uuid });
}

export function resetDefinitionUuid(path: string) {
  return invoke<string>("taskcard_reset_uuid", { path });
}

export function startGroup(prefixPath: string, id: string, sudoPassword?: string) {
  return invoke<void>("taskcard_start_group", {
    prefixPath,
    id,
    sudoPassword: sudoPassword ?? null,
  });
}

export function stopGroup(prefixPath: string, id: string) {
  return invoke<void>("taskcard_stop_group", { prefixPath, id });
}

export function fetchTaskYaml(prefixPath: string, id: string) {
  return invoke<YamlDocument>("taskcard_task_yaml", { prefixPath, id });
}

export function fetchGroupYaml(prefixPath: string, id: string) {
  return invoke<YamlDocument>("taskcard_group_yaml", { prefixPath, id });
}

export function createTaskYaml(content: string, folder: string) {
  return invoke<string>("taskcard_create_task_yaml", { content, folder });
}

export function updateTaskYaml(prefixPath: string, id: string, content: string, folder: string) {
  return invoke<void>("taskcard_update_task_yaml", { prefixPath, id, content, folder });
}

export function deleteTask(prefixPath: string, id: string) {
  return invoke<void>("taskcard_delete_task", { prefixPath, id });
}

export function createGroupYaml(content: string, folder: string) {
  return invoke<string>("taskcard_create_group_yaml", { content, folder });
}

export function updateGroupYaml(prefixPath: string, id: string, content: string, folder: string) {
  return invoke<void>("taskcard_update_group_yaml", { prefixPath, id, content, folder });
}

export function deleteGroup(prefixPath: string, id: string) {
  return invoke<void>("taskcard_delete_group", { prefixPath, id });
}

export function fetchTaskTemplate() {
  return invoke<{ content: string }>("taskcard_task_template");
}

export function fetchGroupTemplate() {
  return invoke<{ content: string }>("taskcard_group_template");
}

export function fetchLogs() {
  return invoke<TaskLogSummary[]>("taskcard_logs");
}

export function fetchHarborLog() {
  return invoke<string>("harbor_self_log");
}

export function readLog(file: string) {
  return invoke<TaskLogContent>("taskcard_read_log", { file });
}

export function readLogChunk(file: string, offset: number, tailLines?: number) {
  return invoke<TaskLogChunk>("taskcard_read_log_chunk", { file, offset, tailLines });
}

export function interfaceUrls(
  task: TaskCardTask,
  settings: Settings | null,
  defaultRouteIp?: string,
  vncPort?: number,
) {
  const workspace = settings?.workspaces.find((item) => item.id === settings.current_workspace);
  const remote = workspace?.mode === "remote";
  const workspaceHost =
    remote && workspace.ssh?.host.trim()
      ? workspace.ssh.host.trim()
      : "127.0.0.1";
  const webviews = (task.webview_interface ?? []).map((panel) => ({
    kind: "webview" as const,
    name: panel.panel_name,
    url: `http://${
      panel.localhost_only ? "127.0.0.1" : defaultRouteIp?.trim() || workspaceHost
    }:${panel.interface_port}/`,
  }));
  const vnc = remote
    ? (task.vnc_interface ?? []).map((panel) => ({
        kind: "vnc" as const,
        name: panel.panel_name,
        url: `http://${workspaceHost}:${vncPort ?? 23682}/`,
      }))
    : [];
  return [...webviews, ...vnc];
}

export function openPanelWindow(title: string, url: string) {
  return invoke<void>("open_panel_window", { title, url });
}

File: src-tauri/crates/harbor_protocol/src/taskcard.rs
use std::collections::HashMap;

use serde::{Deserialize, Deserializer, Serialize};
use uuid::Uuid;

use crate::web_api::{PHYSICAL_VNC_PORT, VIRTUAL_VNC_PORT};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct WebviewInterface {
    pub panel_name: String,
    pub interface_port: u16,
    #[serde(default)]
    pub localhost_only: bool,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct VncInterface {
    pub panel_name: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct TaskConfig {
    pub id: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub env: HashMap<String, String>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct GroupDefinition {
    #[serde(deserialize_with = "deserialize_yaml_version")]
    pub version: String,
    #[serde(default = "generate_uuid")]
    pub uuid: String,
    pub id: String,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub tasks: Vec<GroupTask>,
    #[serde(default)]
    pub folder: String,
    #[serde(default)]
    pub prefix_path: String,
    #[serde(default, skip_deserializing)]
    pub definition_path: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct GroupTask {
    pub task: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub config: String,
    #[serde(default)]
    pub wait_after_sec: u64,
    #[serde(default)]
    pub env: HashMap<String, String>,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub prefix_path: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct TaskSummary {
    pub uuid: String,
    pub uuid_conflict: bool,
    pub id: String,
    pub prefix_path: String,
    pub name: String,
    pub description: String,
    pub workdir: String,
    pub command: String,
    pub env_count: usize,
    pub configs: Vec<TaskConfig>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default_config: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub running_config_id: Option<String>,
    pub requires_sudo: bool,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub webview_interface: Vec<WebviewInterface>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub vnc_interface: Vec<VncInterface>,
    pub folder: String,
    pub status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pid: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub started_at_ms: Option<u128>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub log_file: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TaskCardSnapshot {
    #[serde(default)]
    pub generated_at_ms: u128,
    #[serde(default)]
    pub stale: bool,
    pub root: String,
    #[serde(default = "loopback_address")]
    pub default_route_ip: String,
    #[serde(default = "virtual_vnc_port")]
    pub vnc_port: u16,
    #[serde(default)]
    pub vnc_ready: bool,
    #[serde(default = "physical_vnc_port")]
    pub physical_vnc_port: u16,
    #[serde(default)]
    pub physical_vnc_ready: bool,
    #[serde(default)]
    pub physical_vnc_error: Option<String>,
    pub search_paths: Vec<String>,
    pub discovered_task_dirs: Vec<String>,
    pub discovered_group_dirs: Vec<String>,
    pub tasks: Vec<TaskSummary>,
    pub groups: Vec<GroupDefinition>,
    #[serde(default)]
    pub uuid_conflicts: Vec<UuidConflict>,
    pub errors: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UuidDefinitionRef {
    pub kind: String,
    pub id: String,
    pub prefix_path: String,
    pub path: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UuidConflict {
    pub uuid: String,
    pub definitions: Vec<UuidDefinitionRef>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ResearchResult {
    pub search_paths: Vec<String>,
    pub discovered_task_dirs: Vec<String>,
    pub discovered_group_dirs: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TaskLogSummary {
    pub file: String,
    pub task_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub config_id: Option<String>,
    pub started_at_ms: u128,
    pub modified_at_ms: u128,
    pub bytes: u64,
    pub active: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TaskLogContent {
    pub file: String,
    pub content: String,
    pub truncated: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TaskLogChunk {
    pub file: String,
    pub content: String,
    pub next_offset: u64,
    pub reset: bool,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TaskCardYamlDocument {
    pub content: String,
    pub folder: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ManagedProcessInfo {
    pub pid: i32,
    pub ppid: i32,
    pub name: String,
    pub command: String,
    pub state: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ManagedProcessGroup {
    pub uuid: String,
    pub prefix_path: String,
    pub task_id: String,
    pub leader_pid: u32,
    pub pgid: i32,
    pub started_at_ms: u128,
    pub log_file: String,
    pub config_id: Option<String>,
    pub task_running: bool,
    pub orphaned: bool,
    pub processes: Vec<ManagedProcessInfo>,
}

fn generate_uuid() -> String {
    Uuid::new_v4().to_string()
}

fn loopback_address() -> String {
    "127.0.0.1".into()
}

fn virtual_vnc_port() -> u16 {
    VIRTUAL_VNC_PORT
}

fn physical_vnc_port() -> u16 {
    PHYSICAL_VNC_PORT
}

fn deserialize_yaml_version<'de, D>(deserializer: D) -> Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    struct VersionVisitor;

    impl<'de> serde::de::Visitor<'de> for VersionVisitor {
        type Value = String;

        fn expecting(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            formatter.write_str("a string or number version")
        }

        fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(value.to_string())
        }

        fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(value)
        }

        fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(value.to_string())
        }

        fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(value.to_string())
        }

        fn visit_f64<E>(self, value: f64) -> Result<Self::Value, E>
        where
            E: serde::de::Error,
        {
            Ok(value.to_string())
        }
    }

    deserializer.deserialize_any(VersionVisitor)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn group_contract_accepts_numeric_version_and_generates_uuid() {
        let group: GroupDefinition = serde_yaml::from_str(
            r#"version: 1
id: demo
tasks: []
"#,
        )
        .unwrap();
        assert_eq!(group.version, "1");
        assert!(Uuid::parse_str(&group.uuid).is_ok());
    }

    #[test]
    fn snapshot_defaults_use_protocol_ports() {
        let snapshot: TaskCardSnapshot = serde_json::from_str(
            r#"{
                "root":"/tmp",
                "search_paths":[],
                "discovered_task_dirs":[],
                "discovered_group_dirs":[],
                "tasks":[],
                "groups":[],
                "errors":[]
            }"#,
        )
        .unwrap();
        assert_eq!(snapshot.vnc_port, VIRTUAL_VNC_PORT);
        assert_eq!(snapshot.physical_vnc_port, PHYSICAL_VNC_PORT);
    }
}
